xapian-core/tests/api_sorting.cc

```diff
@@ -321,79 +321,46 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-
+    TEST_EQUAL_DOUBLE(i.get_weight(), new_weight);
+    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), new_weight);
+    TEST_EQUAL_DOUBLE(mymset.get_max_possible(), max(old_max_possible, mymset.get_max_possible()));
     return true;
 }
 
 DEFINE_TESTCASE(replace_weights2, backend) {
-    // open the database (in this case a simple text file
-    // we prepared earlier)
-
     Xapian::Database mydb(get_database("apitest_onedoc"));
-
     Xapian::Enquire enquire(mydb);
-
-    // make a simple query, with one word in it - "word".
     enquire.set_query(Xapian::Query("word"));
-
-    // retrieve the top ten results (we only expect one)
     Xapian::MSet mymset = enquire.get_mset(0, 10);
-
-    // Checking for exception when the number of weights in the input exceeds the number of documents
-    vector<double> weights {-1, -2};
+    vector<double> weights {(mymset.begin()).get_weight(), (mymset.begin()).get_weight()};
     TEST_EXCEPTION(Xapian::InvalidArgumentError, mymset.replace_weights(weights.begin(), weights.end()));
-
     return true;
 }
 
 DEFINE_TESTCASE(sort_existing_mset_by_relevance, backend) {
-    // open the database (in this case a simple text file
-    // we prepared earlier)
     Xapian::Database db = get_database("apitest_simpledata");
     Xapian::Enquire enquire(db);
     enquire.set_query(Xapian::Query("word"));
-
-    // retrieve the top results we expect only two
     Xapian::MSet mymset = enquire.get_mset(0, 10);
-
-    // Storing the docids of the two documents in the original order.
-    vector<Xapian::docid> docids;
-    Xapian::MSetIterator i = mymset.begin();
-    docids.push_back(*i);
-    i++;
-    docids.push_back(*i);
-    // assigning new weights and then re-sorting to reverse the order of relevance
-    vector<double> weights {-2, -1};
+    double old_max_possible = mymset.get_max_possible();
+    double new_weight1 = mymset.get_max_attained() + 0.1;
+    double new_weight2 = mymset.get_max_attained() + 0.2;
+    vector<double> weights {new_weight1, new_weight2};
+    // assigning new weights and resorting to reverse the original order of relevance.
     mymset.replace_weights(weights.begin(), weights.end());
     mymset.sort_by_relevance();
-    int j = -1, k = 1;
-    // checking for the correct order of documents and their respective weights.
-    for (Xapian::MSetIterator m = mymset.begin(); m != mymset.end(); ++m, --j, --k) {
-	TEST_EQUAL_DOUBLE(m.get_weight(), j);
-	TEST_EQUAL(docids[k], *m);
-    }
-    // checking whether the max_attained attribute was updated correctly.
-    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), double(-1));
+    TEST_EQUAL_DOUBLE((mymset.begin()).get_weight(), new_weight2);
+    TEST_EQUAL_DOUBLE((++mymset.begin()).get_weight(), new_weight1);
+    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), max(new_weight1, new_weight2));
+    TEST_EQUAL_DOUBLE(mymset.get_max_possible(), max(old_max_possible, mymset.get_max_possible()));
```

Comment:
Again, without knowing what `old_max_possible` is, you don't know if you're exercising the code correctly. Find out what the actual value will be (by hand) and hard code things higher than it as weights to ensure that it moves. `max()` does not belong in this test!